### PR TITLE
feat: added byte and target token calculations for evaluation datasets, improved calculation logic for BPC/bPPL

### PR DESCRIPTION
**Pull request by FromCSUZhou**
Added byte and target token calculations for evaluation datasets, improved calculation logic for BPC/bPPL

## Summary by CodeRabbit

- **New Features**
  - Added detailed logging of dataset statistics, including total bytes, token counts, and token-to-byte ratio.
  - Introduced calculation and logging of bits-per-character (BPC) and bits-perplexity (bPPL) metrics.
  - Validation now uses BPC as the primary metric for submissions instead of raw loss, with safeguards for invalid or missing data.
  - Added a summary table display for evaluation metrics to improve readability.
  - Implemented functions to compute BPC, bPPL, and token-byte ratio for enhanced evaluation accuracy.

- **Bug Fixes**
  - Added checks to handle invalid evaluation loss values and zero-byte datasets, ensuring robust metric reporting.

### CODE REVIEW

File: src/validate.py [hunk 1]
```diff
@@ -299,6 +300,43 @@ def validate(
         eval_dataset = load_sft_dataset(
             eval_file, context_length, template_name=base_model, tokenizer=tokenizer
         )
+
+        total_bytes = 0
```
Comment (nickcom007) [inline src/validate.py:304]: Please encapsulate the bpc calculation logic in a separate file. maybe `src/core/loss.py`

File: src/validate.py [hunk 2]
```diff
@@ -331,19 +369,74 @@ def validate(
...
-        logger.info("evaluate result is %s" % str(eval_result))
+        logger.info("Raw evaluation result: %s" % str(eval_result))
+        logger.info(f"Average token-level loss (nats): {eval_loss}")
+
         if local_test:
-            logger.info("The model can be correctly validated by validators.")
+            logger.info(
+                "The model can be correctly validated by validators (raw loss)."
+            )
+            # Calculate and log BPC/bPPL even for local tests
```
Comment (nickcom007) [inline src/validate.py:382]: I don't quite understand why we need to repeat all the code here again?

File: src/validate.py [hunk 3]
```diff
@@ -331,19 +369,74 @@ def validate(
...
+            eval_loss_to_submit = LOSS_FOR_MODEL_PARAMS_EXCEED
+        elif (
+            not isinstance(eval_loss, float)
+            or math.isnan(eval_loss)
+            or math.isinf(eval_loss)
         ):
-            eval_loss = LOSS_FOR_MODEL_PARAMS_EXCEED
+            logger.error(f"Invalid eval_loss ({eval_loss}), submitting high loss.")
+            eval_loss_to_submit = LOSS_FOR_MODEL_PARAMS_EXCEED
+
```
[the remark this answers is not in the document]
Comment (nickcom007) [reply src/validate.py:435]: Valid?

**Review by nickcom007 — CHANGES_REQUESTED** [review]

File: src/validate.py [hunk 4]
```diff
@@ -299,6 +300,43 @@ def validate(
...
+            target_ids = [
+                id
+                for id, mask in zip(item["input_ids"], item["target_mask"])
+                if mask == 1
+            ]
+            if target_ids:
+                target_text = tokenizer.decode(target_ids, skip_special_tokens=True)
+                total_bytes += len(target_text.encode("utf-8"))
+                total_target_tokens += len(target_ids)
+
```
[the remark this answers is not in the document]
Comment (FromCSUZhou) [reply src/validate.py:331]: This change inadvertently triggered the Trainer behavior of accumulating all model predictions (logits), which could easily lead to OOM for LLMs.

**Review by nickcom007 — APPROVED** [review]
